### ISSUE TITLE
Add Max Attempt Limit to Notification Rate Limiting

### ISSUE THREAD

**Issue by michaelbarley** [jamesmills/laravel-notification-rate-limit#37]
### Overview
I propose introducing a feature that allows developers to specify a maximum number of attempts for which a notification can be retried before permanently ceasing any further attempts. This feature aims to prevent endless notification retries and enables better control over the notification flow, especially in automated systems such as CRON jobs.

### Problem
Currently, while the Laravel Notification Rate Limit package effectively prevents flooding users with notifications by enforcing a delay between attempts, it does not limit the total number of retries for a notification. In scenarios where notifications are tied to regular checks (e.g., a CRON job checking for low user credits), a notification can be repeatedly sent after the specified delay without a cap on the total attempts. This can potentially lead to situations where notifications become redundant or annoying.

### Example Use Case
Consider a notification system that alerts users about their low credits. A CRON job periodically checks if a user's credits fall below a certain threshold and triggers a notification. The rate limiter is set to delay subsequent notifications for 2 days. However, without a cap on attempts, the user might receive the same notification indefinitely every 2 days if their credit status doesn't change. Ideally, the notification should only be retried a limited number of times, say 3 times, before stopping.

### Proposed Solution
Integrate a setting within the RateLimitedNotification trait that allows specifying a maximum number of delivery attempts. Once a notification reaches this limit, it would not be eligible for further retries. This could be implemented using a counter that increments with each attempt, stored either in the cache or directly managed through Laravel’s existing job mechanism for queued notifications.

### Benefits
- Reduces Redundancy: Ensures notifications do not become repetitive and annoying to users.
- Enhances Control: Gives developers explicit control over the notification lifecycle, allowing them to manage user experience more effectively.
- Promotes Efficiency: Prevents system resources from being wasted on unnecessary notification attempts.

**Comment by tibbsa** [comment]
Thank you for your detailed proposal.

At first blush, I am not sure if the logic for this should be directly a part of this package or not. Storing the counter in the cache is not a permanent solution -- caches may be cleared for any number of reasons, and at that point a user (in your example situation) would start receiving notifications again for a few days before it once again reached the max send count. 

The alternative would be to maintain a table of notifications and the number of time the same notification had been sent to a specific user. However, rather like with another recent suggestion (https://github.com/jamesmills/laravel-notification-rate-limit/issues/33), that involves adding tables to people's databases, which might be suboptimal depending on their performance needs.

My instinct for implementing this in your example scenario would be to add that counter to the user model, increment with every successful send, and then simply stop trying to send notifications to anyone whose count has exceeded the limit. You would reset that to 0 when their credits were replenished.  From a performance standpoint, this might be preferable to adding hundreds or thousands of unnecessary queued "sent credit notification" tasks to the system, only to have this package discard them in any event.  So my first thought is - where the issue is whether the notification should continue to be sent to the user AT ALL, that would be better determined at the application level before it is event sent. 

*However*, as a way of facilitating this in connection with the overall rate limiting logic, I could see adding an optional rate limit "check hook" that would give your application an opportunity to decide for itself whether the notification should be discarded for reasons other than the time-based rate limit.  

You would have a function in your notification similar to the following...

```php
// Return TRUE if this notification should be discarded
public function rateLimitCheckDiscard(string $key): bool
{
   string $counterKey = $key . '_count';

   $count = Cache::get($counterKey, 0);

   if ($count >= 3) {
      return true;
   }

   Cache::put($counterKey, ++$count);

   return false;
}
```

You could then store the counter in cache, in your `$notifiable` (typically `User`) model, or in some other way that fit your needs -- again, remembering that you may need to clear/reset it if the blocking condition eventually clears as well.  

Any thoughts?

**Comment by tibbsa** [comment]
@michaelbarley Just following up on this.  Would this approach be useful to you?

**Comment by michaelbarley** [comment]
@tibbsa thank you for taking the time to provide me  an approach to address my feature suggestion. I like the sound of it!

**Comment by tibbsa** [comment]
Support for this is now available in the v3.0 release. (A major version bump was required because of changes introduced to the discarded notification Event signature.)
